java/arcs/core/data/expression/Expression.kt

```diff
@@ -324,6 +346,80 @@ sealed class Expression<T> {
     class BooleanLiteralExpression(boolean: Boolean) : LiteralExpression<Boolean>(boolean) {
         override fun <Result> accept(visitor: Visitor<Result>) = visitor.visit(this)
     }
+
+    /** Subtypes represent a [Expression]s that operate over the result of the [qualifier]. */
+    interface QualifiedExpression<T> {
+        val qualifier: Expression<Sequence<T>>?
+    }
+
+    /**
+     * Represents an iteration over a [Sequence] in the current scope under the identifier [source],
+     * placing each member of the sequence in a scope under [iterationVar] and evaluating
+     * [iterationExpr], returning a new sequence.
+     *
+     * @param E the type of the [qualifier] expression if any
+     * @param T the type of elements in the resulting [Sequence]
+     */
+    data class FromExpression<E, T>(
+        override val qualifier: Expression<Sequence<E>>?,
```

Comment:
Should FromExpression be a QualifiedExpression? It seems that `qualifier` is null, while it is present in WhereExpression and SelectExpression.

Reply:
Though I guess we could someday allow:
```
from x in [subquery1, subquery2, subquery3]
(...)
```
And here the `qualifier` would be present, but not the `source`. Though you could likely do something similar with other language constructs...

Reply:
You can do this
```
from x in p
from y in p.q
where y.blah == "Hello"
```

In which case, this is 
```
WHERE(FROM(FROM("x", "p"), "y", currentScope["p"]["q"]), currentScope["y"]["blah"] eq "Hello")
```